[ome] Invoke simplifyInstruction after lowering ownership and use replaceAllSimplifiedUsesAndErase instead of a manual RAUW.

gottesmm opened the pull request:
I am doing two things here:

1. simplifyInstruction expects its result to only be passed to
replaceAllSimplifiedUsesAndErase. This was an oversite.

2. I am upstreaming support for InstSimplify in OSSA (to finish sil-combine
work). The main assumption that inst-simplify is going to have is that it is
looking at SIL in correct ownership ssa or non-ownership ssa. It is not going to
be designed to work with an intermediate world that is the world where before
this PR it was being invoked (specifically, we have marked the function as not
being in OSSA but haven't finished transforming it into non-OSSA SIL). To avoid
this problem, I moved the simplification to the end of the pass after we have
done the initial traversal.

Should be NFCI.

gottesmm committed:
[ome] Invoke simplifyInstruction after lowering ownership and use replaceAllSimplifiedUsesAndErase instead of a manual RAUW.

I am doing two things here:

1. simplifyInstruction expects its result to only be passed to
replaceAllSimplifiedUsesAndErase. This was an oversite.

2. I am upstreaming support for InstSimplify in OSSA (to finish sil-combine
work). The main assumption that inst-simplify is going to have is that it is
looking at SIL in correct ownership ssa or non-ownership ssa. It is not going to
be designed to work with an intermediate world that is the world where before
this PR it was being invoked (specifically, we have marked the function as not
being in OSSA but haven't finished transforming it into non-OSSA SIL). To avoid
this problem, I moved the simplification to the end of the pass after we have
done the initial traversal.

Should be NFCI.

gottesmm commented:
@swift-ci smoke test

atrick commented:
Not a big deal, but it's always helpful when renaming is done in a separate commit.

atrick reviewed (approved):
The change in this PR looks good. I asked for some clarification. The only issue I have is how badly SILBuilder is being abused for no reason. Why is the visitor passed a builder at all? The visitor should own a SILBuilderContext. It's trackingList should be that context's InsertedInsts. It can also hold on to the current builder, which can be instantiated in beforeVisit with the visitor's context.

Basically, don't call any setter functions on SILBuilder!

gottesmm commented on lib/SILOptimizer/Mandatory/OwnershipModelEliminator.cpp:
```diff
@@ -45,180 +47,217 @@ DumpBefore("sil-dump-before-ome-to-path", llvm::cl::Hidden);
+      : builder(newBuilder),
+        openedArchetypesTracker(&newBuilder.getFunction()) {
+    newBuilder.setTrackingList(&trackingList);
+    newBuilder.setOpenedArchetypesTracker(&openedArchetypesTracker);
```
I don't know why it was done this way TBH. I am fine changing it to however you want me to change it.

gottesmm commented on lib/SILOptimizer/Mandatory/OwnershipModelEliminator.cpp:
```diff
@@ -45,180 +47,217 @@ DumpBefore("sil-dump-before-ome-to-path", llvm::cl::Hidden);
+    // tracking list that we haven't added yet.
+    drainTrackingList();
+    builder.setInsertionPoint(instToVisit);
+    builder.setCurrentDebugScope(instToVisit->getDebugScope());
```
This is old code. I will fix! Thanks for calling this out!

gottesmm commented on lib/SILOptimizer/Mandatory/OwnershipModelEliminator.cpp:
```diff
@@ -245,158 +284,173 @@ bool OwnershipModelEliminatorVisitor::visitSwitchEnumInst(
 
-      MadeChange |= Visitor.visit(I);
+  // Once we have finished processing all instructions, we should not be
+  // consistently in non-ossa form. Now go through any instructions and simplify
```
@atrick when I was bringing up ossa part of my goal was zero diff (or as small as possible diff) after ownership lowering of our tests. With that in mind, I started to notice that a bunch of diff can from not running inst simplify's simplifications on split up destructures. It was completely a pragmatic decision.

That being said, OME is not an OSSA pass or a non-OSSA pass. It is more of a specialized transform that is in a between world while it is running. I don't think that is an assumption we want any of our utility code to have to deal with. Code should always be able to expect either fully non-ossa or ossa.

So it makes sense to hoist out this simplification code outside of the lowering so that it knows it structurally has non-ossa SIL. The actual thing we are simplifying are struct_extract from splitting destructures during the OME transform. So the simplification has to run after the transform on non-OSSA.

gottesmm replied:
I think I improved the wording.

gottesmm committed:
[ome] Remove bad pattern of having a global SILBuilder with a global SILBuilderWithContext and multiple local SILBuilderWithScope.

To make it a little less verbose, I added an always inline helper method called:

template <typename ResultTy>
ResultTy OME::withBuilder(SILInstruction *insertPt, function_ref<ResultTy (SILBuilder &, SILLocation)> visitor) {
  SILBuilderWithScope builder(insertPt, builderContext);
  return visitor(builder, insertPt->getLoc());
}

Since it is always inline there shouldn't be any perf impact and it should be
like invoking a lambda in the caller directly in the same function that the
lambda was declared in. I just couldn't type SILBuilderWithScope
builder(insertPt, builderContext) all the time. Seems error prone.

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
I am going to rebase this PR on top of #34844. #34844 fixes the bug that this PR was hitting (as one can see from the failing tests).

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci benchmark

gottesmm commented:
@swift-ci test linux platform

gottesmm commented:
@swift-ci test

gottesmm commented:
Woot all tests!

atrick commented on lib/SILOptimizer/Mandatory/OwnershipModelEliminator.cpp:
```diff
@@ -57,16 +57,35 @@ namespace {
+  withBuilder(SILInstruction *insertPt,
+              llvm::function_ref<ResultTy(SILBuilder &, SILLocation)> visitor) {
+    SILBuilderWithScope builder(insertPt, builderCtx);
+    return visitor(builder, insertPt->getLoc());
```
It's too bad this is so awkward. Ideally, building a single instruction should be able to naturally use function chaining:

`SILBuilder(inst, builderCtx).emitSomethingElse(operands)`

gottesmm replied:
Yes. But today it is SILBuilderWithScope.

atrick reviewed (approved):
Thanks for fixing SILBuilder

gottesmm commented:
Going to rebase again.

gottesmm commented:
Going to rebase on top of #34877

gottesmm commented:
@swift-ci test

gottesmm commented:
Rebased on top of main now that I landed the precursors.

gottesmm commented:
@swift-ci test
